packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,561 @@
+    try {
+      return (await this.messagingSystem.call('SnapController:handleRequest', {
+        snapId: snapId as SnapId,
+        origin: 'metamask',
+        handler: HandlerType.OnAssetsLookup,
+        request: {
+          jsonrpc: '2.0',
+          method: 'onAssetLookup',
+          params: {
+            assets,
+          },
+        },
+      })) as Promise<AssetMetadataResponse>;
+    } catch (error) {
+      // Ignore
+      console.error(error);
+      return undefined;
+    }
+  }
+
+  /**
+   * Get assets list for an account
+   *
+   * @param accountId - AccountId to get assets for
+   * @param snapId - Snap ID for the account
+   * @returns list of assets
+   */
+  async #getAssetsList(
+    accountId: string,
+    snapId: string,
+  ): Promise<CaipAssetTypeOrId[]> {
+    return await this.#getClient(snapId).listAccountAssets(accountId);
+  }
+
+  /**
+   * Gets a `KeyringClient` for a Snap.
+   *
+   * @param snapId - ID of the Snap to get the client for.
+   * @returns A `KeyringClient` for the Snap.
+   */
+  #getClient(snapId: string): KeyringClient {
+    return new KeyringClient({
+      send: async (request: JsonRpcRequest) =>
+        (await this.messagingSystem.call('SnapController:handleRequest', {
+          snapId: snapId as SnapId,
+          origin: 'metamask',
+          handler: HandlerType.OnKeyringRequest,
+          request,
+        })) as Promise<Json>,
+    });
+  }
+
+  /**
+   * Assert that the controller mutex is locked.
+   *
+   * @throws If the controller mutex is not locked.
+   */
+  #assertControllerMutexIsLocked() {
+    if (!this.#controllerOperationMutex.isLocked()) {
+      throw new Error(
+        'MultichainAssetsControllerError - Attempt to update state',
+      );
+    }
+  }
+
+  /**
+   * Lock the controller mutex before executing the given function,
+   * and release it after the function is resolved or after an
+   * error is thrown.
+   *
+   * This wrapper ensures that each mutable operation that interacts with the
+   * controller and that changes its state is executed in a mutually exclusive way,
+   * preventing unsafe concurrent access that could lead to unpredictable behavior.
+   *
+   * @param callback - The function to execute while the controller mutex is locked.
+   * @returns The result of the function.
+   */
+  async #withControllerLock<Result>(
+    callback: MutuallyExclusiveCallback<Result>,
+  ): Promise<Result> {
+    return withLock(this.#controllerOperationMutex, callback);
+  }
+}
+
+/**
+ * Lock the given mutex before executing the given function,
+ * and release it after the function is resolved or after an
+ * error is thrown.
+ *
+ * @param mutex - The mutex to lock.
+ * @param callback - The function to execute while the mutex is locked.
+ * @returns The result of the function.
+ */
+async function withLock<Result>(
+  mutex: Mutex,
+  callback: MutuallyExclusiveCallback<Result>,
+): Promise<Result> {
+  const releaseLock = await mutex.acquire();
```

Comment:
It looks like the `async-mutex` package has a `mutex.runExclusive` function, would that also work here instead of this `withLock` function? https://github.com/DirtyHairy/async-mutex/blob/8e38062a7f20ddf30143c4bce23ef8a672136ed6/src/Semaphore.ts#L39 (this is in the Semaphore class, but it seems Mutex uses Semaphore under the hood)

Reply:
We wanted to use the same pattern used in the `keyringController`; looking at `async-mutex` it does looks like
` await mutex.acquire()` will call Semaphore under the hood; so if its fine maybe we can keep same pattern?

Reply:
Oh okay, no problem then.